### PR TITLE
Handle loose type parameters in Extract Method

### PR DESCRIPTION
**Pull request by amcasey**
Known limitations:

1. If a type parameter on an inner symbol shadows a type parameter on an
outer symbol, the generated code will be incorrect.  We should either
rename one or more type parameters or forbid the extraction.

2. Type arguments are always passed explicitly, even if they would be
inferred correctly.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -476,6 +476,38 @@ namespace ts {
         return false;
     }
 
+    /* @internal */
+    export function isDeclarationWithTypeParameters(node: Node): node is DeclarationWithTypeParameters;
+    export function isDeclarationWithTypeParameters(node: DeclarationWithTypeParameters): node is DeclarationWithTypeParameters {
```
Comment (minestarks) [inline src/compiler/utilities.ts]: >export function isDeclarationWithTypeParameters(node: DeclarationWithTypeParameters): node is DeclarationWithTypeParameters { [](start = 4, length = 125)

What is the purpose of the overload here? I think you can just remove it and have one function that takes Node. 

If you have an argument of DeclarationWithTypeParameters type then you already know enough at compile time not to call this function.

Comment (minestarks) [inline src/compiler/utilities.ts]: Or is this a trick to get the `default: staticAssertNever` to work?

---
In reply to: [135145388](https://github.com/Microsoft/TypeScript/pull/17988#discussion_r135145388) [](ancestors = 135145388)

File: src/services/refactors/extractMethod.ts [hunk 2]
```diff
@@ -921,7 +1022,42 @@ namespace ts.refactor.extractMethod {
...
+                if (hasTypeParameters(node)) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        function recordTypeParameterUsages(type: Type) {
+            // PERF: This is potentially very expensive.  `type` could be a library type with
```
Comment (minestarks) [inline src/services/refactors/extractMethod.ts]: >This is potentially very expensive [](start = 21, length = 34)

Is this a good point to check for cancellation then?

File: src/compiler/utilities.ts (hunk 1 above)
Comment (amcasey) [inline src/compiler/utilities.ts]: It's the tricky one. 😄

**Comment by minestarks** [comment]
> 

Would it be cheap to just disallow this case?

---
Refers to: tests/baselines/reference/extractMethod/extractMethod14.ts:40 in 8dd757a. [](commit_id = 8dd757a6dad5382767141aaa0d2dbc55cd57b314, deletion_comment = False)

**Comment by minestarks** [comment]
:shipit:

File: src/services/refactors/extractMethod.ts (hunk 2 above)
Comment (amcasey) [reply src/services/refactors/extractMethod.ts]: I think cancellation would need to be pushed down into the walker to make it useful.  That's a big enough change that I'd prefer to make it separately.

**Comment by minestarks** [comment]
I'm extremely confused by the mouse emoji I apparently just posted, but I meant to approve this in Codeflow.

**Pull request merged**